src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,305 @@
+
+  Image
+  (to-clojure [this _]
+    {:tag   (node-to-tag this)
+     :attrs {:src (str (.getUrl this))
+             :alt (str (.getText this))
+             :title (not-empty (str (.getTitle this)))}})
+
+  AutoLink
+  (to-clojure [this _]
+    {:tag   (node-to-tag this)
+     :attrs {:href (str (.getUrl this))}})
+
+  MailLink
+  (to-clojure [this _]
+    {:tag   (node-to-tag this)
+     :attrs {:address (str (.getText this))}})
+
+  HtmlEntity
+  (to-clojure [this _]
+    {:tag (node-to-tag this)
+     :content (str (.getChars this))})
+
+  HtmlBlock
+  (to-clojure [this _]
+    (str (.getChars this)))
+
+  HtmlInline
+  (to-clojure [this _]
+    (str (.getChars this)))
+
+  HtmlCommentBlock
+  (to-clojure [this _]
+    (str (.getChars this)))
+
+  HtmlInlineComment
+  (to-clojure [this _]
+    (str (.getChars this)))
+
+  nil
+  (to-clojure [this _]
+    nil))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                        Slack markup generation                                                 |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(def ^:private html-entities
+  (edn/read-string (slurp (io/resource "html-entities.edn"))))
+
+(def ^:private escaped-chars-regex
+  #"\\[\\/*_`'\[\](){}<>#+-.!$@%^&=|\?~]")
+
+(defn- escape-text
+  "Insert zero-width characters before and after certain characters that are escaped in the Markdown
+  (or are otherwise parsed as plain text) to prevent them from being parsed as formatting in Slack."
+  [string]
+  (-> string
+      ;; First, remove backslashes from escaped formatting characters since they're not removed during Markdown parsing
+      (str/replace escaped-chars-regex #(str (second %1)))
+      ;; Add a soft hyphen around certain chars to avoid triggering formatting in Slack
+      (str/replace "&" "\u00ad&\u00ad")
+      (str/replace ">" "\u00ad>\u00ad")
+      (str/replace "<" "\u00ad<\u00ad")
+      (str/replace "*" "\u00ad*\u00ad")
+      (str/replace "_" "\u00ad_\u00ad")
+      (str/replace "`" "\u00ad`\u00ad")
+      (str/replace "~" "\u00ad~\u00ad")))
+
+(defn- resolve-uri
+  "If the provided URI is a relative path, resolve it relative to the site URL so that links work
+  correctly in Slack/Email."
+  [uri]
+  (when uri
+    (if-let [site-url (public-settings/site-url)]
+      (.toString (.resolve (new URI ^String site-url) ^String uri))
+      uri)))
+
+(defn- ast->mrkdwn
+  "Takes an AST representing Markdown input, and converts it to a mrkdwn string that will render nicely in Slack.
+
+  Some of the differences to Markdown include:
+    * All headers are just rendered as bold text.
+    * Ordered and unordered lists are printed in plain text.
+    * Inline images are rendered as text that links to the image source, e.g. <image.png|[Image: alt-text]>."
+  [{:keys [tag attrs content]}]
```

Comment:
One thing about this makes me nervous. Repeated concatenation of large strings can get really bad for performance. Is it possible to change to a `StringBuilder`?

```clojure
qp=> (time
  (do (reduce str "" (range 10000))
      nil))
"Elapsed time: 46.697234 msecs"
nil
qp=> (time
  (do (str (reduce (fn [^StringBuilder sb x] (.append sb (str x))) (StringBuilder.) (range 10000)))
      nil))
"Elapsed time: 1.524527 msecs"
nil
```

Reply:
That's an interesting point. I'd be happy to try that, but my main concern is that some of the more complex string manipulations might be difficult, or at least much less readable, using `StringBuilder`. For example, I'm not sure how to directly translate the logic for `:unordered-list-item` or `:ordered-list-item`. Also, since this is a recursive tree traversal, there might be a number of `StringBuilder`s being created for the leaves that end up being joined together -- not sure how this affects the efficiency.

Ultimately, most markdown cards won't have more than a few levels of nested elements, and simple text cards will be returned as strings directly with no concatenation involved, so this might be premature optimization?